### PR TITLE
[Container app] `az containerapp env create/update`: Support peer-to-peer traffic encryption with `--enable-peer-to-peer-encryption`

### PR DESCRIPTION
**Pull request by LGDoor**
---

This checklist is used to make sure that common guidelines for a pull request are followed.

### Related command
az containerapp env create/update

### General Guidelines

- [ ] Have you run `azdev style <YOUR_EXT>` locally? (`pip install azdev` required)
- [ ] Have you run `python scripts/ci/test_index.py -q` locally? (`pip install wheel==0.30.0` required)
- [ ] My extension version conforms to the [Extension version schema](https://github.com/Azure/azure-cli/blob/release/doc/extensions/versioning_guidelines.md)

For new extensions:

- [ ] My extension description/summary conforms to the [Extension Summary Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/extensions/extension_summary_guidelines.md).

### About Extension Publish

There is a pipeline to automatically build, upload and publish extension wheels.  
Once your pull request is merged into main branch, a new pull request will be created to update `src/index.json` automatically.  
You only need to update the version information in file setup.py and historical information in file HISTORY.rst in your PR but do not modify `src/index.json`.

### CODE REVIEW

**Comment by yonzhan** [comment]
Thank you for your contribution! We will review the pull request and get back to you soon.

**Comment by LGDoor** [comment]
pending on 2024-02-02-preview

**Comment by LGDoor** [comment]
@Greedygre @Juliehzl plz review

File: src/containerapp/azext_containerapp/containerapp_env_decorator.py [hunk 1]
```diff
@@ -149,6 +148,21 @@ def set_up_custom_domain_configuration(self):
...
+        is_p2p_encryption_enabled = self.get_argument_p2p_encryption_enabled()
+        is_mtls_enabled = self.get_argument_mtls_enabled()
+        logger.info("mtls: p2p_encryption: %s, mtls: %s", is_p2p_encryption_enabled, is_mtls_enabled)
+
+        if is_p2p_encryption_enabled is not None:
+            safe_set(self.managed_env_def, "properties", "peerTrafficConfiguration", "encryption", "enabled", value=is_p2p_encryption_enabled)
+            if is_p2p_encryption_enabled is False:  # keep the two properties consistent.
+                safe_set(self.managed_env_def, "properties", "peerAuthentication", "mtls", "enabled", value=False)
+
+        if is_mtls_enabled is not None:
```
Comment (Greedygre) [inline src/containerapp/azext_containerapp/containerapp_env_decorator.py:161]: Do we accept `--enable-peer-to-peer-encryption --enable-mtls` with different value?

**Comment by Greedygre** [comment]
please get approval from your team who could be responsible for your business logic.

**Comment by Greedygre** [comment]
The CI integration Failed is due to the azure-cli has upgraded api-version to 2024-03-01, I will update the recording files soon.

File: src/containerapp/azext_containerapp/containerapp_env_decorator.py (hunk 1 above)
Comment (LGDoor) [reply src/containerapp/azext_containerapp/containerapp_env_decorator.py:161]: Yes. `--enable-peer-to-peer-encryption true` and `--enable-mtls false` is a valid combination.

**Comment by LGDoor** [comment]
@zhenqxuMSFT  plz help review

**Comment by Greedygre** [comment]
Recommend title:
`az containerapp env create/update`: Support peer-to-peer traffic encryption with `--enable-peer-to-peer-encryption`

File: src/containerapp/azext_containerapp/_params.py [hunk 2]
```diff
@@ -344,3 +344,6 @@ def load_arguments(self, _):
         c.argument('environment_name', options_list=['--environment'], help="The environment name.")
         c.argument('resource_group_name', arg_type=resource_group_name_type, id_part=None)
         c.argument('configuration', nargs="*", help="Java component configuration. Configuration must be in format \"<propertyName>=<value> <propertyName>=<value> ...\".")
+
+    with self.argument_context('containerapp env', arg_group='Peer Traffic Configuration') as c:
+        c.argument('p2p_encryption_enabled', arg_type=get_three_state_flag(), options_list=['--enable-peer-to-peer-encryption'], help='Boolean indicating whether the peer-to-peer traffic encryption is enabled for the environment.')
```
Comment (Greedygre) [inline src/containerapp/azext_containerapp/_params.py:349]: This params only exist in extension, it should add `is_preview=True`

Comment (Greedygre) [inline src/containerapp/azext_containerapp/_params.py:349]: The param's name is pretty long, recommend giving a shorter name too.

File: src/containerapp/azext_containerapp/containerapp_env_decorator.py [hunk 3]
```diff
@@ -149,6 +148,21 @@ def set_up_custom_domain_configuration(self):
...
 
+    def set_up_peer_to_peer_encryption(self):
+        is_p2p_encryption_enabled = self.get_argument_p2p_encryption_enabled()
+        is_mtls_enabled = self.get_argument_mtls_enabled()
+        logger.info("mtls: p2p_encryption: %s, mtls: %s", is_p2p_encryption_enabled, is_mtls_enabled)
+
+        if is_p2p_encryption_enabled is not None:
+            safe_set(self.managed_env_def, "properties", "peerTrafficConfiguration", "encryption", "enabled", value=is_p2p_encryption_enabled)
+            if is_p2p_encryption_enabled is False:  # keep the two properties consistent.
+                safe_set(self.managed_env_def, "properties", "peerAuthentication", "mtls", "enabled", value=False)
```
Comment (Greedygre) [inline src/containerapp/azext_containerapp/containerapp_env_decorator.py:159]: The logic here means if `--enable-peer-to-peer-encryption False`, set:
`properties.peerTrafficConfiguration.encryption.enabled` as `False`
`properties.peerAuthentication.mtls.enabled` as `False`

But what if `--enable-peer-to-peer-encryption False --enable-mtls`?
`properties.peerAuthentication.mtls.enabled` will be set as `False`
Then in line 162 be set to True.
But 
`properties.peerTrafficConfiguration.encryption.enabled` will be set as `True` at line 164

Comment (Greedygre) [reply src/containerapp/azext_containerapp/containerapp_env_decorator.py:159]: The logic you add to `keep the two properties consistent.` make some input not make sense

File: src/containerapp/azext_containerapp/containerapp_env_decorator.py [hunk 4]
```diff
@@ -217,12 +239,29 @@ def set_up_custom_domain_configuration(self):
...
+
+        if is_mtls_enabled is not None:
+            safe_set(self.managed_env_def, "properties", "peerAuthentication", "mtls", "enabled", value=is_mtls_enabled)
+            if is_mtls_enabled is True:  # keep the two properties consistent.
+                safe_set(self.managed_env_def, "properties", "peerTrafficConfiguration", "encryption", "enabled", value=True)
```
Comment (Greedygre) [inline src/containerapp/azext_containerapp/containerapp_env_decorator.py:255]: Same as above

File: src/containerapp/azext_containerapp/containerapp_env_decorator.py (hunk 1 above)
Comment (Greedygre) [reply src/containerapp/azext_containerapp/containerapp_env_decorator.py:161]: @LGDoor 
But what if --enable-peer-to-peer-encryption False --enable-mtls?
What is the result?

File: src/containerapp/azext_containerapp/containerapp_env_decorator.py [hunk 5]
```diff
@@ -149,6 +148,21 @@ def set_up_custom_domain_configuration(self):
                 }
             self.managed_env_def["properties"]["customDomainConfiguration"] = custom_domain
 
+    def set_up_peer_to_peer_encryption(self):
+        is_p2p_encryption_enabled = self.get_argument_p2p_encryption_enabled()
+        is_mtls_enabled = self.get_argument_mtls_enabled()
+        logger.info("mtls: p2p_encryption: %s, mtls: %s", is_p2p_encryption_enabled, is_mtls_enabled)
```
Comment (zhenqxuMSFT) [inline src/containerapp/azext_containerapp/containerapp_env_decorator.py:154]: Is this log visible to customer too?

File: src/containerapp/azext_containerapp/containerapp_env_decorator.py (hunk 3 above)
Comment (zhenqxuMSFT) [reply src/containerapp/azext_containerapp/containerapp_env_decorator.py:159]: I think you should write a warning logs to customer if the customer provided value is overwritten, or fail the command in validation logic.

**Review by Moazzem-Hossain-pixel — CHANGES_REQUESTED** [review]
Pls bug fixes

File: src/containerapp/azext_containerapp/_params.py (hunk 2 above)
Comment (LGDoor) [reply src/containerapp/azext_containerapp/_params.py]: This parameter name is already signed off by PM. We have to keep it and exclude the warning from the linter.

File: src/containerapp/azext_containerapp/containerapp_env_decorator.py (hunk 5 above)
Comment (LGDoor) [reply src/containerapp/azext_containerapp/containerapp_env_decorator.py]: You're right. It should be removed.

File: src/containerapp/azext_containerapp/containerapp_env_decorator.py (hunk 1 above)
Comment (LGDoor) [reply src/containerapp/azext_containerapp/containerapp_env_decorator.py:162]: Added validation to fail this combination.

**Review by Moazzem-Hossain-pixel — CHANGES_REQUESTED** [review]
Pls do as well.

File: src/containerapp/azext_containerapp/_params.py (hunk 2 above)
Comment (Greedygre) [reply src/containerapp/azext_containerapp/_params.py]: I mean to can provide an additional short name.

File: src/containerapp/azext_containerapp/containerapp_env_decorator.py [hunk 6]
```diff
@@ -58,13 +57,17 @@ def validate_arguments(self):
...
-        
+
         # validate custom domain configuration
         if self.get_argument_hostname():
             if self.get_argument_certificate_file() and self.get_argument_certificate_key_vault_url():
                 raise ValidationError("Cannot use --certificate-file with --certificate-akv-url at the same time")
             if (not self.get_argument_certificate_file()) and (not self.get_argument_certificate_key_vault_url()):
                 raise ValidationError("Either --certificate-file or --certificate-akv-url should be set when --dns-suffix is set")
+            
+        # validate mtls and p2p traffic encryption
```
Comment (Greedygre) [inline src/containerapp/azext_containerapp/containerapp_env_decorator.py:68]: Please add test to cover your new params and the logic

**Review by zhenqxuMSFT — APPROVED** [review]

**Review by Greedygre — APPROVED** [review]

**Review by zhoxing-ms — APPROVED** [review]

**Pull request merged**